devolutions-gateway/src/http/controllers/sogar_token.rs

```diff
@@ -0,0 +1,94 @@
+use crate::config::{Config, SogarUser};
+use picky::{
+    jose::{jws::JwsAlg, jwt::JwtSig},
+    key::PrivateKey,
+};
+use saphir::{
+    controller::Controller,
+    http::{Method, StatusCode},
+    macros::controller,
+    prelude::Request,
+};
+use serde::{Deserialize, Serialize};
+use slog_scope::error;
+use sogar_core::AccessToken;
+use std::sync::Arc;
+
+pub struct TokenController {
+    config: Arc<Config>,
+}
+
+impl TokenController {
+    pub fn new(config: Arc<Config>) -> Self {
+        Self { config }
+    }
+}
+
+#[controller(name = "registry")]
+impl TokenController {
+    #[post("/oauth2/token")]
+    async fn get_token(&self, mut req: Request) -> (StatusCode, Option<String>) {
+        match req.form::<AccessToken>().await {
+            Ok(body) => {
+                let password_out = body.password.clone();
+                let username_out = body.username;
+
+                let config = self.config.clone();
+
+                for user in &config.sogar_user {
+                    if let (Some(username), Some(hashed_password)) = (&user.username, &user.password) {
+                        if username == &username_out {
+                            let matched = argon2::verify_encoded(hashed_password.as_str(), password_out.as_bytes());
+                            if matched.is_err() || !matched.unwrap() {
+                                return (StatusCode::UNAUTHORIZED, None);
+                            }
+
+                            return create_token(&config.delegation_private_key, user);
```

Comment:
You don't need that `clone`
```suggestion
                for user in &self.config.sogar_user {
                    if let (Some(username), Some(hashed_password)) = (&user.username, &user.password) {
                        if username == &username_out {
                            let matched = argon2::verify_encoded(hashed_password.as_str(), password_out.as_bytes());
                            if matched.is_err() || !matched.unwrap() {
                                return (StatusCode::UNAUTHORIZED, None);
                            }

                            return create_token(&self.config.delegation_private_key, user);
```

Reply:
done